physical/consul.go

```diff
@@ -577,3 +360,270 @@ func (c *ConsulLock) Value() (bool, string, error) {
 	return held, value, nil
 }
+
+func (c *ConsulBackend) NotifyActiveStateChange() error {
+	select {
+	case c.notifyActiveCh <- notifyEvent{}:
+	default:
+		// NOTE: If this occurs Vault's active status could be out of
+		// sync with Consul until reconcileTimer expires.
+		c.logger.Printf("[WARN]: consul: Concurrent state change notify dropped")
+	}
+
+	return nil
+}
+
+func (c *ConsulBackend) NotifySealedStateChange() error {
+	select {
+	case c.notifySealedCh <- notifyEvent{}:
+	default:
+		// NOTE: If this occurs Vault's sealed status could be out of
+		// sync with Consul until checkTimer expires.
+		c.logger.Printf("[WARN]: consul: Concurrent sealed state change notify dropped")
+	}
+
+	return nil
+}
+
+func (c *ConsulBackend) checkDuration() time.Duration {
+	return lib.DurationMinusBuffer(c.checkTimeout, checkMinBuffer, checkJitterFactor)
+}
+
+func (c *ConsulBackend) RunServiceDiscovery(shutdownCh ShutdownChannel, advertiseAddr string, activeFunc activeFunction, sealedFunc sealedFunction) (err error) {
+	if err := c.setAdvertiseAddr(advertiseAddr); err != nil {
+		return err
+	}
+
+	go c.runEventDemuxer(shutdownCh, advertiseAddr, activeFunc, sealedFunc)
+
+	return nil
+}
+
+func (c *ConsulBackend) runEventDemuxer(shutdownCh ShutdownChannel, advertiseAddr string, activeFunc activeFunction, sealedFunc sealedFunction) {
+	// Fire the reconcileTimer immediately upon starting the event demuxer
+	reconcileTimer := time.NewTimer(0)
+	defer reconcileTimer.Stop()
+
+	// Schedule the first check.  Consul TTL checks are passing by
+	// default, checkTimer does not need to be run immediately.
+	checkTimer := time.NewTimer(c.checkDuration())
+	defer checkTimer.Stop()
+
+	// Use a reactor pattern to handle and dispatch events to singleton
+	// goroutine handlers for execution.  It is not acceptable to drop
+	// inbound events from Notify*().
+	//
+	// goroutines are dispatched if the demuxer can acquire a lock (via
+	// an atomic CAS incr) on the handler.  Handlers are responsible for
+	// deregistering themselves (atomic CAS decr).  Handlers and the
+	// demuxer share a lock to synchronize information at the beginning
+	// and end of a handler's life (or after a handler wakes up from
+	// sleeping during a back-off/retry).
+	var shutdown bool
+	var checkLock int64
+	var registeredServiceID string
+	var serviceRegLock int64
+shutdown:
+	for {
+		select {
+		case <-c.notifyActiveCh:
+			// Run reconcile immediately upon active state change notification
+			reconcileTimer.Reset(0)
+		case <-c.notifySealedCh:
+			// Run check timer immediately upon a seal state change notification
+			checkTimer.Reset(0)
+		case <-reconcileTimer.C:
+			// Unconditionally rearm the reconcileTimer
+			reconcileTimer.Reset(reconcileTimeout - lib.RandomStagger(reconcileTimeout/checkJitterFactor))
+
+			// Abort if service discovery is disabled or a
+			// reconcile handler is already active
+			if !c.disableRegistration && atomic.CompareAndSwapInt64(&serviceRegLock, 0, 1) {
```

Comment:
The original patch did that.  I pushed the disableRegistration down to the handlers so that if we wanted to change this at runtime via a HUP there would be less code to chase.  In the back of my mind it seemed plausible that someone would start with `disabled_registration=true` and then would want to migrate to a Vault managed Consul service world.  Work to support changing this at runtime isn't there, but by pushing it further down into the runloop it makes that future ask easier.  I'm completely okay pushing this check back into `RunServiceDiscovery()` and short circuiting the demuxer entirely.